### DISCUSSION TITLE
How to detect if a worker process has crashed?

### DISCUSSION THREAD

**Discussion by HenrikBengtsson** [r-lib/mirai#108] [Q&A] [answered]
What happens if a worker crashes, i.e. the R process abruptly terminates? Will there be a timeout somewhere?

For example,

```r
library(mirai)
daemons(2L, dispatcher = FALSE)

m <- mirai({ tools::pskill(pid = Sys.getpid()) })
unresolved(m)
#> [1] TRUE
```

Now, calling:

```r
call_mirai(m)
```

hangs because there's no response. However, will this eventually timeout, or will it wait forever?  Is it possible to detect that the underlying worker for this `mirai` object has crashed before calling `call_mirai()`?

FYI, it also looks like `call_mirai()` cannot be interrupted here, e.g. Ctrl-C does nothing.

**Comment by shikokuchuo** [comment, marked as answer]
For non-dispatcher daemons, you probably want:
```r
daemons(2L, dispatcher = FALSE, resilience = FALSE)
```
in which case 'm' resolves to an errorValue 19.

For interruptible call_mirai, use `call_mirai_()`.

**Reply by HenrikBengtsson** [reply]
I see. In my example, which uses the default `resilience = TRUE`, does **mirai** keep trying to process that expression over and over in new processes?

**Reply by shikokuchuo** [reply]
`status()` is your friend here. You'll see that in your case, connections will be 0. i.e. both daemons will have been killed. In such a case there is little you can do but reset. mirai relies on NNG req/rep protocol's automatic retry system and it will retry on other daemons if evaluation was not successful on one. There is no automatic launching of new daemons.

For the dispatcher case, retries are isolated to each daemon instance. This means that your expression above won't bring down all daemons, but it will be retried on all new instances, hence causing those to fail. In such a case, you can use `saisei(force = TRUE)` to rotate that instance and the mirai will be returned as an errorValue, and a new daemon instance will be able to connect.
